### PR TITLE
map integration

### PR DESCRIPTION
**Pull request by kolkri**
I added a Map component to show the map and the markers as locations of Saunas:
image

Everything works fine but I get 2 warnings I can not solve:

1) With typescript it complains _id (in line 54) type because that is how it is in MongoDB:
image

2) I get this warning in console about how each child should have unique key prop, I am not sure is this because of the mapping (picture above line 52), and how to add key there when usually it is put inside of a html tag?
image

I also changed the fetching data through /api/saunas, like discussed last time. I also used that in the map component, I don't know if my way is the best, any comments on this? 
image
image

### CODE REVIEW

File: sauna-app/components/map.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+
+
+function Map() {
+
+  const [saunas, setSaunas] = useState<SaunaData>();
+
+  useEffect(() => {
+    const fetchAPI = async () => {
+      await fetch('http://localhost:3000/api/saunas')
```
Comment (eevajonnapanula) [inline sauna-app/components/map.tsx]: A couple of things: You'll want to use `/api/saunas` instead of with the localhost, because it won't work once deployed with localhost. It's enough, because you're working inside Next. 

You also want to move fetching saunas to `index.tsx` and then just pass the saunas-array down to `Map`-component. You don't want to fetch any data in `components`-folder's components; only in the `pages`.

File: sauna-app/pages/index.tsx [hunk 2]
```diff
@@ -26,12 +34,22 @@ function SaunaPage(props: SaunaData) {
...
+  //const saunas = await fetchData()
+  
+  //OPTION 2 TO FETCH DATA USING API
+  const saunas: SaunaData = await fetch('http://localhost:3000/api/saunas')
+    .then((response) => {
+      return response.json();
+    })
+    .then((data) => {
+      return data
+    })
```
Comment (eevajonnapanula) [inline sauna-app/pages/index.tsx:45]: You'll want to go with the second one.

File: sauna-app/pages/index.tsx [hunk 3]
```diff
@@ -26,12 +34,22 @@ function SaunaPage(props: SaunaData) {
...
 
 //data fetching for pre-rendering
-export async function getStaticProps() {
-  const saunas = await fetchData()
+export async function getStaticProps(): Promise<SaunaProps> {
+  //OPTION 1 TO FETCH DATA:
+  //const saunas = await fetchData()
+  
+  //OPTION 2 TO FETCH DATA USING API
+  const saunas: SaunaData = await fetch('http://localhost:3000/api/saunas')
```
Comment (eevajonnapanula) [inline sauna-app/pages/index.tsx:39]: A note about `async`/`await` and promise callbacks; don't mix them. I know these can feel confusing, and it takes time with them, so no worries 😊 

A quick explanation and how this can be solved:
1. You can use `async`/`await`:
```js
const saunasResponse: SaunaRespData = await fetch('/api/saunas')

const saunas: SaunasData = saunasResponse.data.saunas
```

or

```js
const saunas: SaunaData = await fetch('/api/saunas')
.then((response) => {
      return response.json();
    })
    .then((data) => {
      return data
    })
```

File: sauna-app/pages/index.tsx [hunk 4]
```diff
@@ -26,12 +34,22 @@ function SaunaPage(props: SaunaData) {
...
+      return response.json();
+    })
+    .then((data) => {
+      return data
+    })
 
   return {
     props: {
-      saunas: saunas.map(s => ({
+      saunas: saunas.saunas.map(s => ({
```
Comment (eevajonnapanula) [inline sauna-app/pages/index.tsx:49]: Nit: You could maybe destructure the object to have `_id` and `rest`, and then spread the rest-object, and then set the id. I'm not giving (yet) an example here; This is extra if you want  to study more. This code works, and what I'm suggesting here is more for learning :) 

So, two things to google: Object desctucturing (with rest-parameter) and object spreading. And if you have any questions, don't hesitate to ask 😊

File: sauna-app/pages/typeDefinitions.tsx [hunk 5]
```diff
@@ -1,4 +1,4 @@
-interface SingleSauna {
+export interface SingleSauna {
```
Comment (eevajonnapanula) [inline sauna-app/pages/typeDefinitions.tsx]: For solving the TypeScript error with `_id`, here's one possible solution (note: Naming of the types is not the best, I'm only on my first cup of coffee this morning 😅 ): 

Have a common type for all the common properties, and then subtype for the Sauna used in the app, and Sauna from the API:

```ts
interface SaunaProperties {
  // common props
}

interface SaunaResp extends SaunaProperties {
  _id: string
}

interface SaunaInApp extends SaunaProperties {
  id: string
}

type SingleSauna = SaunaInApp | SaunaResp
```

This way, either `id` or `_id` is mandatory property (but not both).

File: sauna-app/components/map.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    <MapContainer className={style.map} center={[60.257215, 24.966394]} zoom={10} scrollWheelZoom={false}>
+      <TileLayer
+        attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
+        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
+      />
+      {saunas?.saunas.map(s => {
+        let latitude: number = s.latitude
+        let longitude: number = s.longitude
+        let name: string = s.name
+        let id: string =s.id
```
Comment (eevajonnapanula) [inline sauna-app/components/map.tsx]: Nit: Object destructuring could be used here too.

File: sauna-app/components/map.tsx [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+          </Marker>
+        )
+      })
+
+      }
+    </MapContainer>
+  )
+} 
+
+let DefaultIcon = L.icon({
```
Comment (eevajonnapanula) [inline sauna-app/components/map.tsx]: Nit: You could use `const` instead of `let` here as it doesn't change. Overall I'd recommend using `const` in most cases - usually we don't want to change the value of the variable, but assign a new one if something changes.

File: sauna-app/components/map.tsx [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    fetchAPI()
+  }, [])
+
+
+  return (
+    <MapContainer className={style.map} center={[60.257215, 24.966394]} zoom={10} scrollWheelZoom={false}>
+      <TileLayer
+        attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
+        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
```
Comment (eevajonnapanula) [inline sauna-app/components/map.tsx:14]: A question: Where this url comes from and what does it do? Because it seems that is should be a string literal with variables `s`, `x`, `y` and `z` 🤔

File: sauna-app/components/map.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        let latitude: number = s.latitude
+        let longitude: number = s.longitude
+        let name: string = s.name
+        let id: string =s.id
+        return (
+          <Marker 
+            key={id}
+            position={[latitude, longitude]}
+            eventHandlers={{
+              mouseover: (event) => event.target.openPopup(),
```
Comment (eevajonnapanula) [inline sauna-app/components/map.tsx]: From accessibility perspective, you probably want to add a focus event listener too to open the popup. But I'm fine if that's done in a separate PR :)

**Review by eevajonnapanula — CHANGES_REQUESTED** [review]
Overall looks good, and I love how the map looks! 

I've left some comments; the ones starting with "Nit" are things that you can decide if you want to change them, the ones without it are the ones I'd like to see addressed. 

And for the key-error; that's an odd one, because I can see that there's a key-prop in the `Marker`-component, which is what it's probably complaining about as it's the only mapped thing there. Can you log if the id's happen to be same for different saunas?

File: sauna-app/components/map.tsx (hunk 8 above)
Comment (kolkri) [reply sauna-app/components/map.tsx:14]: Hmm not exactly sure about that url, I just copy pasted that from React Leaflet component, see this example: [https://react-leaflet.js.org/](https://react-leaflet.js.org/)

File: sauna-app/components/map.tsx (hunk 9 above)
Comment (kolkri) [reply sauna-app/components/map.tsx]: I actually changed the Popup tags to Tooltip tags because they have the hovering effects on them so less code, but I have to read/test more how they work with focusing.

File: sauna-app/pages/typeDefinitions.tsx (hunk 5 above)
Comment (kolkri) [reply sauna-app/pages/typeDefinitions.tsx]: I was struggling with this a lot and somehow I could not make it work, I now used maybe a bad solution to use only _id (instead of id) everywhere to get rid of the warnings, I don't know is this a bad way to do it or does it really matter how I have named them (_id/id) ?

File: sauna-app/pages/index.tsx (hunk 4 above)
Comment (kolkri) [reply sauna-app/pages/index.tsx:49]: I will look into this, and if I don't get it work maybe we can discuss about this in our next meeting :)

File: sauna-app/pages/index.tsx (hunk 3 above)
Comment (kolkri) [reply sauna-app/pages/index.tsx:39]: I tried to replace `http://localhost:3000/api/saunas` to `/api/saunas` in the getStaticProps() function in the index.tsx file. For some reason it does not word, after reading some documentation I think it could be maybe something to do with the pre-rendering functionalities. But after I deploy the project could I just change the fetching address to `DEPLOYED_URL/api/saunas` ?

**Comment by kolkri** [comment]
I made the changes, so asking re-review for this branch. I had some issues, see the comments. :)

File: sauna-app/components/map.tsx (hunk 8 above)
Comment (eevajonnapanula) [reply sauna-app/components/map.tsx:14]: It seems that there's an explanation: [https://leafletjs.com/reference.html#tilelayer](https://leafletjs.com/reference.html#tilelayer) so I guess it's okay as is 😄 I have worked with Leaflet before, but it's a long time ago.

File: sauna-app/pages/index.tsx (hunk 3 above)
Comment (eevajonnapanula) [reply sauna-app/pages/index.tsx:39]: I started digging deeper into this, and I'm super sorry for giving incorrect information before; so apparently, you can actually do what you did (= import the `loadSaunas`-function and use it in `getStaticProps`). But if you do that, I'd put the load function under `lib/load-saunas`  or something - not in the `api`-routes. 

Here's more information about that: https://nextjs.org/docs/basic-features/data-fetching/get-static-props#write-server-side-code-directly

So, you can forget that `localhost`-comment, and if you want, you can either merge this PR with the `localhost` and create another one to revert back to the `load`-function usage (and move it to `lib` (or utils or whatever you want to call it) -folder), or you can do it in this PR. Up to you 😊

File: sauna-app/pages/index.tsx (hunk 4 above)
Comment (eevajonnapanula) [reply sauna-app/pages/index.tsx:49]: Sounds like a plan!

File: sauna-app/pages/typeDefinitions.tsx (hunk 5 above)
Comment (eevajonnapanula) [reply sauna-app/pages/typeDefinitions.tsx]: I don't know if it's a bad one - it's a solution. There might be some people who say that it's unconventional, but then again, I think everything is just about agreements and how we set the conventions. Some use underscore to annotate private fields (esp. with classes). 

So I'm okay to accept this solution, just a note that when you're interviewing with some company, they might ask about this so be prepared to explain why you ended up with this solution. (My answer would be something in the lines of "I was lazy and wanted to use the same interface for the response and in the components, and did not want to add extra code just so that I could rename one key")

**Review by eevajonnapanula — APPROVED** [review]
Looks good! There's that one comment about `getStaticProps` and loading saunas, which is something that would be good to be refactored at some point, but not a requirement for merging this PR :)
